Mypyc 1.13.0 crashes on specific match statement

<!--
  Use this form only if mypy reports an "INTERNAL ERROR" and/or gives a traceback.
  Please include the traceback and all other messages below (use `mypy --show-traceback`).
-->

**Crash Report**

First time trying to compile my Python project with Mypyc; it seems to hang itself up on a specific `match` statement. Although there are a couple more in the code, it only complains about this one as far as I can tell. Replacing it with an semantically equivalent if-else gives me type warnings and errors instead (the expected result considering that these seem all valid).

**Traceback**

```
❯ uv run --with mypy[mypyc] mypyc -p nyl --show-traceback
Traceback (most recent call last):
  File "mypyc/irbuild/builder.py", line 272, in accept
  File "mypy/nodes.py", line 1648, in accept
  File "mypyc/irbuild/visitor.py", line 252, in visit_match_stmt
  File "mypyc/irbuild/statement.py", line 1022, in transform_match_stmt
  File "mypy/nodes.py", line 1648, in accept
  File "mypyc/irbuild/match.py", line 92, in visit_match_stmt
  File "mypy/patterns.py", line 150, in accept
  File "mypyc/irbuild/match.py", line 164, in visit_class_pattern
src/nyl/generator/components.py:81: AssertionError: 
```

**To Reproduce**

1. Clone https://github.com/helsing-ai/nyl and check out `0.8.1` (that is HEAD today)
2. Run `uv run --with mypy[mypyc]==1.13.0 mypyc -p nyl`

The specific piece of code that it seems to hiccup on is this:

https://github.com/helsing-ai/nyl/blob/659ffec8e4b8362f8012e0b096fe4b2606861586/src/nyl/generator/components.py#L81-L92

```py
        match component:
            case HelmComponent(path):
                chart = HelmChart(
                    metadata=instance.metadata,
                    spec=HelmChartSpec(
                        chart=ChartRef(path=str(path.resolve())),
                        values={"metadata": resource["metadata"], **instance.spec},
                    ),
                )
                return self.helm_generator.generate(chart)
            case _:
                raise RuntimeError(f"unexpected component type: {component}")
```

There are a couple more match statements in the code which it doesn't seem to complain about after changing this to use if-else instead. I'm not exactly sure where best to start to reduce it to a smaller example.

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 1.13.0
- Mypy command-line flags: -p nyl
- Mypy configuration options from `mypy.ini` (and other config files):
    ```
    explicit_package_bases = true
    namespace_packages = true
    show_column_numbers = true
    strict = true
    mypy_path = ["src"]
    ```
- Python version used: 3.11.9
- Operating system and version: WSL2 Ubuntu 22.04.5 LTS

--- a/mypyc/irbuild/match.py
+++ b/mypyc/irbuild/match.py
@@ -16,7 +16,7 @@ from mypy.patterns import (
     ValuePattern,
 )
 from mypy.traverser import TraverserVisitor
-from mypy.types import Instance, TupleType, get_proper_type
+from mypy.types import Instance, LiteralType, TupleType, get_proper_type
 from mypyc.ir.ops import BasicBlock, Value
 from mypyc.ir.rtypes import object_rprimitive
 from mypyc.irbuild.builder import IRBuilder
@@ -152,23 +152,7 @@ class MatchVisitor(TraverserVisitor):
 
             node = pattern.class_ref.node
             assert isinstance(node, TypeInfo)
-
-            ty = node.names.get("__match_args__")
-            assert ty
-
-            match_args_type = get_proper_type(ty.type)
-            assert isinstance(match_args_type, TupleType)
-
-            match_args: list[str] = []
-
-            for item in match_args_type.items:
-                proper_item = get_proper_type(item)
-                assert isinstance(proper_item, Instance) and proper_item.last_known_value
-
-                match_arg = proper_item.last_known_value.value
-                assert isinstance(match_arg, str)
-
-                match_args.append(match_arg)
+            match_args = extract_dunder_match_args_names(node)
 
             for i, expr in enumerate(pattern.positionals):
                 self.builder.activate_block(self.code_block)
@@ -355,3 +339,24 @@ def prep_sequence_pattern(
             patterns.append(pattern)
 
     return star_index, capture, patterns
+
+
+def extract_dunder_match_args_names(info: TypeInfo) -> list[str]:
+    ty = info.names.get("__match_args__")
+    assert ty
+    match_args_type = get_proper_type(ty.type)
+    assert isinstance(match_args_type, TupleType)
+
+    match_args: list[str] = []
+    for item in match_args_type.items:
+        proper_item = get_proper_type(item)
+
+        match_arg = None
+        if isinstance(proper_item, Instance) and proper_item.last_known_value:
+            match_arg = proper_item.last_known_value.value
+        elif isinstance(proper_item, LiteralType):
+            match_arg = proper_item.value
+        assert isinstance(match_arg, str), f"Unrecognized __match_args__ item: {item}"
+
+        match_args.append(match_arg)
+    return match_args

--- a/mypyc/test-data/irbuild-match.test
+++ b/mypyc/test-data/irbuild-match.test
@@ -1727,3 +1727,81 @@ L4:
 L5:
 L6:
     unreachable
+
+[case testMatchLiteralMatchArgs_python3_10]
+from typing_extensions import Literal
+
+class Foo:
+    __match_args__: tuple[Literal["foo"]] = ("foo",)
+    foo: str
+
+def f(x: Foo) -> None:
+    match x:
+        case Foo(foo):
+            print("foo")
+        case _:
+            assert False, "Unreachable"
+[out]
+def Foo.__mypyc_defaults_setup(__mypyc_self__):
+    __mypyc_self__ :: __main__.Foo
+    r0 :: str
+    r1 :: tuple[str]
+L0:
+    r0 = 'foo'
+    r1 = (r0)
+    __mypyc_self__.__match_args__ = r1
+    return 1
+def f(x):
+    x :: __main__.Foo
+    r0 :: object
+    r1 :: i32
+    r2 :: bit
+    r3 :: bool
+    r4 :: str
+    r5 :: object
+    r6, foo, r7 :: str
+    r8 :: object
+    r9 :: str
+    r10 :: object
+    r11 :: object[1]
+    r12 :: object_ptr
+    r13, r14 :: object
+    r15 :: i32
+    r16 :: bit
+    r17, r18 :: bool
+L0:
+    r0 = __main__.Foo :: type
+    r1 = PyObject_IsInstance(x, r0)
+    r2 = r1 >= 0 :: signed
+    r3 = truncate r1: i32 to builtins.bool
+    if r3 goto L1 else goto L3 :: bool
+L1:
+    r4 = 'foo'
+    r5 = CPyObject_GetAttr(x, r4)
+    r6 = cast(str, r5)
+    foo = r6
+L2:
+    r7 = 'foo'
+    r8 = builtins :: module
+    r9 = 'print'
+    r10 = CPyObject_GetAttr(r8, r9)
+    r11 = [r7]
+    r12 = load_address r11
+    r13 = PyObject_Vectorcall(r10, r12, 1, 0)
+    keep_alive r7
+    goto L8
+L3:
+L4:
+    r14 = box(bool, 0)
+    r15 = PyObject_IsTrue(r14)
+    r16 = r15 >= 0 :: signed
+    r17 = truncate r15: i32 to builtins.bool
+    if r17 goto L6 else goto L5 :: bool
+L5:
+    r18 = raise AssertionError('Unreachable')
+    unreachable
+L6:
+    goto L8
+L7:
+L8:
+    return 1

